File: javascript/beer-song/beer-song.js
var BeerSong = function(){};

const bottlesOnWall = ' bottles of beer on the wall';
const bottles = ' bottles of beer.\n';
const bottleOnWall = ' bottle of beer on the wall';
const bottle = ' bottle of beer.\n';
const takeDown = 'Take one down and pass it around, ';

BeerSong.prototype.verse = function(verseNum){
  const pluralCurrentWall = verseNum + bottlesOnWall + ', ' + verseNum + bottles;
  const pluralUpdatedWall = takeDown + (verseNum - 1) + bottlesOnWall + '.\n';
  const newlySingleUpdatedWall = takeDown + (verseNum - 1) + bottleOnWall + '.\n';
  const singularCurrentWall = verseNum + bottleOnWall + ', ' + verseNum + bottle;
  const singularUpdatedWall =
    'Take it down and pass it around, no more bottles of beer on the wall.\n';

  if (verseNum > 2) {
    return pluralCurrentWall + pluralUpdatedWall;
  } else if (verseNum === 2) {
    return pluralCurrentWall + newlySingleUpdatedWall;
  } else if (verseNum === 1){
    return singularCurrentWall + singularUpdatedWall;
  } else {
    return 'No more' + bottlesOnWall + ', no more' + bottles +
      'Go to the store and buy some more, 99' + bottlesOnWall + '.\n';
  }
};

BeerSong.prototype.sing = function(startVerse, endVerse = 0) {
  let verses = "";
  for (var i = startVerse; i >= endVerse; i-- ) {
    verses += this.verse(i);
    if (i !== endVerse) {
      verses += '\n';
    }
  }
  return verses;
};

module.exports = BeerSong;

File: ruby/beer-song/beer_song.rb
class BeerSong
  def initialize
    @num_bottles = 0
  end

  def verse(verse_number)
    @num_bottles = verse_number

    num_bottles == 0 ? no_more_bottles : bottles_on_wall + take_down
  end

  def verses(starting_verse, ending_verse)
    output = ''
    starting_verse.downto(ending_verse) do |verse_number|
      output += "#{verse(verse_number)}\n"
    end
    output.chomp
  end

  def lyrics
    verses(99, 0)
  end

  private

  attr_reader :num_bottles

  def bottles_on_wall
    noun = num_bottles > 1 ? 'bottles' : 'bottle'

    "#{num_bottles} #{noun} of beer on the wall, " \
    "#{num_bottles} #{noun} of beer.\n"
  end

  def take_down
    noun = num_bottles > 1 ? 'one' : 'it'

    "Take #{noun} down and pass it around, " \
    "#{remaining_bottles} of beer on the wall.\n"
  end

  def no_more_bottles
    "No more bottles of beer on the wall, no more bottles of beer.\n" \
    "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
  end

  def remaining_bottles
    case num_bottles
    when 1
      'no more bottles'
    when 2
      '1 bottle'
    else
      "#{num_bottles - 1} bottles"
    end
  end
end

module BookKeeping
  VERSION = 2
end
